### ISSUE TITLE
how paf solve split brain problem when pacemaker is dead

### ISSUE THREAD

**Issue by huxiang44** [ClusterLabs/PAF#153]
when pacemaker is dead, paf will detect, right?
paf will recognize primary is down, and failover another standby. But the old primary is normal.
Other standbys still connect old primary, and transactions can commit on it.
But after the one failovered become primary, it will hang, right?
how paf solve above problem

**Comment by ioguix** [comment]
Hello,

PAF is just a resource agent. It does not check for Pacemaker health. It's just a glue between Pacemaker and PostgreSQL.

If Pacemaker fails on one node, either the watchdog (if configured) will self-kill the node or other nodes in the cluster will fence it using your fencing setup.

Having said that, I should admit I did not understand your scenario clearly. Could you rephrase it?

Regards,

**Comment by huxiang44** [comment]
Sorry, I maybe misunderstand the role of paf and pacemaker. 
So, if the paf is fake dead ( that may because network isolation between paf and peacemarker), the pecemarker will kill paf and the corresponding primary. But if priamry is not down immediately, write transactions can be committed on it. Once new primary is up, it may hang if old primary is not down yet ( because it can not write enough log on standbys). But if old priamry is down and standbys connect to new primary, they will find diverged logs and should rewind them. As a result, the transactions committed during the period are lost.

**Comment by ioguix** [comment]
Hi,

> So, if the paf is fake dead

PAF can not be dead, it's a script (but it can have bugs). The network, server, PostgreSQL or other component could be dead.

> the pecemarker will kill paf and the corresponding primary

Pacemaker nodes will take a decision to kill the dead node and one of the safe node in your cluster will trigger the fencing command you setup.

> But if priamry is not down immediately

Yes it is. The fencing command is supposed to either kill the server or its access to the data immediately before reporting its success to the cluster.

> Once new primary is up, it may hang if old primary is not down yet ( because it can not write enough log on standbys)

I'm not sure to understand. Maybe you are using synchronous replication?  But anyway, the old master must be down after the fencing command, a new master is then promoted and your standby shoudl reconnect to the new master.

> But if old priamry is down and standbys connect to new primary, they will find diverged logs and should rewind them. 

No. PAF takes care to promote the most advanced remaining standby in replication, so other standby can replicate from the new master with no trouble.

> As a result, the transactions committed during the period are lost.

depend on your replication setup. You will lost no transaction in synchronous rep.

**Comment by ioguix** [comment]
Please, make sure to read the various documentation about PostgreSQL replication, Pacemaker architecture and mechanism, then PAF before building your cluster.

Building an proper HA architecture is hard and require a lot of time and team investment. Admins should be aware of the stack and be able to quickly manage it and react after an automatic failover to get back on a safe situation as fast as possible (eg. next morning).

You and your team need to learn and train on a regular basis to keep the stack clean. If you do not, the SPoF of your stack will be your own team.

**Comment by huxiang44** [comment]
Thanks very much for giving explanation in details. From my understanding, the most importance point is to guarantee the old primary is killed before promoting another standby.  From the start, i suspect that killing primary may fail in some cases, so there may have two primary and the old one is still working during that period. But as you have showed the fencing command can assure it.

**Comment by huxiang44** [comment]
Absolutely, i should read docs for more details.

**Comment by huxiang44** [comment]
Alsolutely , i should read more docs on it.  Regards!

**Comment by ioguix** [comment]
Hi,

Should the fencing command fails or timeout, Pacemaker will cancel all other tasks and try to takes new decisions. So no promotion will happen until the failing master is known stopped or dead from Pacemaker point of view. No matter how many time Pacemaker will try stopping or kill the the old master.

In fact, after some amount of failing transition, by default it just give up and leave the cluster with no master at all.

Regards,
